### ISSUE TITLE
hs_model_program - migration problem when using www resources

### ISSUE THREAD

**Issue by mjstealey** [hydroshare/hydroshare#892]
@mohamedmorsy, @Castronova - Could you have a look at the console output for the migration of the build using the develop branch with www resources: http://ci.hydroshare.org:8080/view/2.%20Deployment/job/branch-builder/22/console

This is also deployed on https://riverdev.hydroshare.org if you'd like to check or test any resources in a deployed environment.

Error observed:
```bash
Operations to perform:
  Synchronize unmigrated apps: debug_toolbar, galleries, irods_browser_app, accounts, conf, rest_framework, forms, corsheaders, generic, boot, comments, blog, jquery_ui, haystack, crispy_forms, core, hs_metrics, filebrowser_safe, sitemaps, gis, pages, mobile, autocomplete_light, djcelery, widget_tweaks, django_extensions
  Apply all migrations: redirects, theme, hs_geo_raster_resource, hs_geographic_feature_resource, ga_resources, ga_ows, hs_labels, hs_model_program, sites, contenttypes, hs_tools_resource, hs_access_control, hs_app_netCDF, hs_modelinstance, ref_ts, sessions, hs_swat_modelinstance, auth, django_docker_processes, hs_app_timeseries, oauth2_provider, admin, hs_core, django_irods
Synchronizing apps without migrations:
  Creating tables...
  Installing custom SQL...
  Installing indexes...
Running migrations:
  Applying hs_model_program.0006_auto_20151216_1500...Traceback (most recent call last):
  File "manage.py", line 8, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 161, in handle
    executor.migrate(targets, plan, fake=options.get("fake", False))
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 68, in migrate
    self.apply_migration(migration, fake=fake)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 102, in apply_migration
    migration.apply(project_state, schema_editor)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/migration.py", line 108, in apply
    operation.database_forwards(self.app_label, schema_editor, project_state, new_state)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/operations/models.py", line 282, in database_forwards
    getattr(new_model._meta, self.option_name, set()),
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/schema.py", line 310, in alter_unique_together
    ", ".join(columns),
ValueError: Found wrong number (0) of constraints for hs_model_program_mpmetadata(content_type_id, object_id)
```

You can get a copy of the www resources to use locally [here](http://distribution.hydroshare.org/public_html/.data/)

**Comment by mohamedmorsy** [comment]
@mjstealey I talked with @Castronova  about that. This problem was not there before as the Model Program is is currently deployed and working on www.hydroshare.org.  Could this have something to do with existing (old) resources?  @hyi  you did help us before to fix the migrations problems when we moved from ModelProgram version 1 to version 2. Do you think that this problem is related to the old Model Program resources?

**Comment by mjstealey** [comment]
> Do you think that this problem is related to the old Model Program resources

This could definitely be due to old resources if the migration components do not adjust old resources to comply with your newer schema.

The error looks to be related to a wrong number of constraints which could mean it's colliding with an older resource definition.

@hyi and @pkdash had to do something similar when we migrated away from using django for access control and shifted to the hs_access database. @pkdash had a migration that he adjusted to account for the new schema item in older resources. In this case the "resource" in question was user account.

**Comment by hyi** [comment]
@mohamedmorsy @mjstealey @pkdash This looks like a new migration issue different from the migration issues we ran into when mp is upgraded to version 2, but looks like a conflict with existing resources that caused this migration issue as well. Last time due to big changes in mp metadata in version 2, we had to delete all old mp and its dependent resources and recreate them on the latest version, then I have to modify 6 or so Django tables manually in order to preserve the uuids of two old mp and mi resources. Hopefully we don't have to go through the manual process this time around.
